src/go/guestagent/pkg/procnet/scanner_linux.go

```diff
@@ -0,0 +1,237 @@
+// removes DNAT rules that allow traffic to be forwarded to the specified port on localhost, based on
+// the provided 'action' ('append' or 'delete').
+//
+// The function iterates over the provided list of port bindings and, for each binding where the HostIP
+// is set to 127.0.0.1, it constructs and executes the corresponding iptables command to either add or
+// delete the appropriate DNAT rule.
+//
+// The iptables rule ensures that incoming traffic from outside  of the network namespace (i.e.,
+// from the host windows machine) on the specified port and protocol is redirected to the same
+// port on localhost, where the container's service can be accessed.
+//
+// Example iptables rule when 'action' is "append":
+// iptables -t nat -A PREROUTING -p tcp --dport 8009 -j DNAT --to-destination 127.0.0.1:8009
+//
+// Example iptables rule when 'action' is "delete":
+// iptables -t nat -D PREROUTING -p tcp --dport 8009 -j DNAT --to-destination 127.0.0.1:8009
+func (p *ProcNetScanner) execLoopbackIPtablesRule(bindings []nat.PortBinding, portProto nat.Port, action action) error {
+	for _, binding := range bindings {
+		if binding.HostIP == "127.0.0.1" {
+			// iptables -t nat -D PREROUTING -p tcp --dport 8009 -j DNAT --to-destination 127.0.0.1:8009
+			iptablesCmd := exec.CommandContext(p.context,
+				"iptables",
+				"--table", "nat",
+				fmt.Sprintf("--%s", action), "PREROUTING",
+				"--protocol", portProto.Proto(),
+				"--dport", binding.HostPort,
+				"--jump", "DNAT",
+				"--to-destination", fmt.Sprintf("%s:%s", binding.HostIP, binding.HostPort),
+			)
+			if err := iptablesCmd.Run(); err != nil {
+				return err
+			}
+			log.Debugf("running the following iptables rule [%s] for port bindings: %v", iptablesCmd.String(), binding)
+		}
+	}
+	return nil
+}
+
+func addValidProtoEntryToPortMap(entry procnettcp.Entry, portMap nat.PortMap) error {
+	switch entry.Kind {
+	case procnettcp.TCP:
+		if entry.State == procnettcp.TCPListen {
+			return addEntryToPortMap(entry, portMap)
+		}
+	case procnettcp.UDP:
+		if entry.State == procnettcp.UDPEstablished {
+			return addEntryToPortMap(entry, portMap)
+		}
+	}
+	return nil
+}
+
+func addEntryToPortMap(entry procnettcp.Entry, portMap nat.PortMap) error {
+	port := strconv.Itoa(int(entry.Port))
+	portMapKey, err := nat.NewPort(entry.Kind, port)
+	if err != nil {
+		return fmt.Errorf("generating portMapKey protocol: %s, port: %d failed: %w",
+			entry.Kind,
+			entry.Port,
+			err)
+	}
+
+	// It's important not to use entry.IP directly here, as any IP
+	// other than 127.0.0.1 (localhost) or 0.0.0.0 may not be accessible
+	// from the host. To ensure consistent behavior, we always set the
+	// HostIP to INADDR_ANY (0.0.0.0) unless the IP is localhost or 0.0.0.0.
+	// The API tracker will then adjust the address as necessary:
+	// - If admin privileges are enabled, the address will remain 0.0.0.0.
+	// - Otherwise, it will be changed to 127.0.0.1 to ensure proper local binding.
+	var hostIP string
```

Comment:
It may be easier to leave `hostIP` as a `net.IP` and call `.String()` when assigning it to `portBinding`.
```golang
hostIP := net.IPv4(0, 0, 0, 0)
if entry.IP.IsLoopback() {
  hostIP = entry.IP // do we need to worry about `127.1.2.3`?
}
portBinding := nat.PortBinding{
  HostIP:   hostIP.String()
  HostPort: port,
}
```